src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts

```diff
@@ -0,0 +1,122 @@
+﻿import { orderBy, orderByDesc } from './sortingHelper'
+
+export {
+    all,
+    any,
+    clear,
+    distinct,
+    firstOrDefault,
+    forEach,
+    lastOrDefault,
+    max,
+    min,
+    orderBy,
+    orderByDesc,
+    remove,
+    removeFirst
+}
+
+function any<T>(source: T[], predicate: (s: T) => boolean): boolean {
+    return firstOrDefault(source, predicate) != null;
+}
+
+function all<T>(source: T[], predicate: (s: T) => boolean) {
+    for (let i = 0; i < source.length; i++) {
+        if (!predicate(source[i])) {
+            return false;
+        }
+    }
+    return true;
+}
+
+function clear<T>(source: T[]): void {
+    source.splice(0, source.length);
+}
+
+function distinct<T>(source: T[]): T[] {
+    let r = [];
+    for (let i = 0; i < source.length; i++) {
+        let found = false;
+        for (var j = 0; j < r.length; j++) {
+            if (r[j] == source[i]) {
+                found = true;
+                break;
+            }
+        }
+        if (found)
+            continue;
+        r.push(source[i]);
+    }
+    return r;
+}
+
+function firstOrDefault<T>(source: T[], predicate: (s: T) => boolean): T | null {
+    for (let i = 0; i < source.length; i++) {
+        if (predicate(source[i])) {
+            return source[i];
+        }
+    }
+    return null;
+}
+
+function forEach<T>(items: T[], action: (s: T, i: number) => void): void {
+    for (let i = 0; i < items.length; i++) {
+        action(items[i], i);
+    }
+}
+
+function lastOrDefault<T>(source: T[], predicate: (s: T) => boolean): T | null {
+    let lastSatisfyingElement = null;
+    for (let i = 0; i < source.length; i++) {
+        if (predicate(source[i])) {
+            lastSatisfyingElement = source[i];
```

Comment:
Just from a performance standpoint, isn't it better to reverse the `for` statement and iterate backward and return the first result?

Reply:
Yes it is. I copied the implementation from `firstOrDefault` and did not realize I can easily optimize this. I will push a fix for this